secret-provider/src/main/scala/io/lenses/connect/secrets/cache/TtlCache.scala

```diff
@@ -0,0 +1,45 @@
+package io.lenses.connect.secrets.cache
+
+import com.typesafe.scalalogging.LazyLogging
+
+import java.time.Clock
+import scala.collection.concurrent.TrieMap
+
+class TtlCache[V](implicit clock: Clock) extends LazyLogging {
+
+  private val cache = TrieMap[String, ValueWithTtl[V]]()
+
+  private def put(
+    key:   String,
+    value: ValueWithTtl[V],
+  ): ValueWithTtl[V] = {
+    cache.put(key, value)
+    value
+  }
+
+  private def get(
+    key:          String,
+    fnConditions: Seq[ValueWithTtl[V] => Boolean] = Seq(_ => true),
+  ): Option[ValueWithTtl[V]] =
+    cache.get(key).filter(value => fnConditions.forall(_(value)))
+
+  def cachingWithTtl(
+    key:                String,
+    fnCondition:        V => Boolean                       = _ => true,
+    fnGetMissingValue:  () => Either[Throwable, ValueWithTtl[V]],
+    fnFilterReturnKeys: ValueWithTtl[V] => ValueWithTtl[V] = e => e,
+  ): Either[Throwable, ValueWithTtl[V]] =
+    getCachedValue(key, fnCondition)
+      .orElse(fnGetMissingValue().map(put(key, _)))
+      .map(fnFilterReturnKeys)
+
+  private def getCachedValue(key: String, fnCondition: V => Boolean): Either[Throwable, ValueWithTtl[V]] =
```

Comment:
It might be better to merge get and getCachedValue. There is little thing `get` does extra. TrieMap returns Option but in this case, `getCachedValue` returns Either. It's best to keep the Option and let the caller site handle None. The wrapping of Either here does not add much value. I do not know how it compiles since `get` returns Option and then the code does `get(..).toEither` yet the return type is Either[T, ValueWithTtl[V]]; I would have expected EIther[T, Option[ValueWithTtl[V]]].